BUG: 1.22.4 breaks frombuffer API compared with 1.22.3

### Describe the issue:

Arrays returned by np.frombuffer() do not have the same base attribute when compared before and after version 1.22.4. I believe this behavior may only be reproduceable on Windows/OSX.

Related:
- https://github.com/numpy/numpy/pull/21324
- https://github.com/tomopy/tomopy/pull/585

### Reproduce the code example:

```python
import numpy as np
import multiprocessing as mp
import ctypes
shared_obj = mp.RawArray(ctypes.c_double, (6, 6))
shared_arr = np.frombuffer(shared_obj)
assert type(shared_arr) == np.ndarray
assert type(shared_obj) == type(shared_arr.base)
```


### Error message:

```shell
>       assert type(shared_obj) == type(shared_arr.base)
E       AssertionError: assert <class 'multiprocessing.sharedctypes.c_double_Array_2'> == <class 'memoryview'>
E        +  where <class 'multiprocessing.sharedctypes.c_double_Array_2'> = type(<multiprocessing.sharedctypes.c_double_Array_2 object at 0x0000010E0B283940>)
E        +  and   <class 'memoryview'> = type(<memory at 0x0000010E0921ED00>)
E        +    where <memory at 0x0000010E0921ED00> = array([6., 6.]).base
```


### NumPy/Python version information:

The code example works on 1.22.3, but not 1.22.4

--- a/numpy/core/src/multiarray/ctors.c
+++ b/numpy/core/src/multiarray/ctors.c
@@ -3687,15 +3687,16 @@ PyArray_FromBuffer(PyObject *buf, PyArray_Descr *type,
     }
 
     /*
-     * The array check is probably unnecessary.  It preserves the base for
-     * arrays.  This is the "old" buffer protocol, which had no release logic.
-     * (It was assumed that the result is always a view.)
-     *
-     * NOTE: We could also check if `bf_releasebuffer` is defined which should
-     *       be the most precise and safe thing to do.  But that should only be
-     *       necessary if unexpected backcompat issues are found downstream.
+     * If the object supports `releasebuffer`, the new buffer protocol allows
+     * tying the memories lifetime to the `Py_buffer view`.
+     * NumPy cannot hold on to the view itself (it is not an object) so it
+     * has to wrap the original object in a Python `memoryview` which deals
+     * with the lifetime management for us.
+     * For backwards compatibility of `arr.base` we try to avoid this when
+     * possible.  (For example, NumPy arrays will never get wrapped here!)
      */
-    if (!PyArray_Check(buf)) {
+    if (Py_TYPE(buf)->tp_as_buffer
+            && Py_TYPE(buf)->tp_as_buffer->bf_releasebuffer) {
         buf = PyMemoryView_FromObject(buf);
         if (buf == NULL) {
             return NULL;

--- a/numpy/core/tests/test_multiarray.py
+++ b/numpy/core/tests/test_multiarray.py
@@ -5608,12 +5608,13 @@ class TestFromBuffer:
         buf = x.tobytes()
         assert_array_equal(np.frombuffer(buf, dtype=dt), x.flat)
 
-    def test_array_base(self):
-        arr = np.arange(10)
-        new = np.frombuffer(arr)
-        # We currently special case arrays to ensure they are used as a base.
-        # This could probably be changed (removing the test).
-        assert new.base is arr
+    @pytest.mark.parametrize("obj", [np.arange(10), b"12345678"])
+    def test_array_base(self, obj):
+        # Objects (including NumPy arrays), which do not use the
+        # `release_buffer` slot should be directly used as a base object.
+        # See also gh-21612
+        new = np.frombuffer(obj)
+        assert new.base is obj
 
     def test_empty(self):
         assert_array_equal(np.frombuffer(b''), np.array([]))

